pkgs/build-support/docker/tarsum.nix

```diff
@@ -0,0 +1,36 @@
+{ stdenv, go, docker }:
+
+stdenv.mkDerivation {
+  name = "tarsum";
+
+  nativeBuildInputs = [ go ];
+  disallowedReferences = [ go ];
+
+  dontUnpack = true;
+
+  CGO_ENABLED = 0;
+  GOFLAGS = "-trimpath";
+  GO111MODULE = "off";
+
+  buildPhase = ''
+    runHook preBuild
+    mkdir tarsum
+    cd tarsum
+    cp ${./tarsum.go} tarsum.go
+    export GOPATH=$(pwd)
+    export GOCACHE="$TMPDIR/go-cache"
+    mkdir -p src/github.com/docker/docker/pkg
+    ln -sT ${docker.moby-src}/pkg/tarsum src/github.com/docker/docker/pkg/tarsum
```

Comment:
Symlinking the source isn't really compatible with `buildGo{Module,Package}` so I'll leave it as it is for now.

We'll likely need to change this to use go modules and version it when `go` deprecates `GO111MODULE = off` but I don't see much benefit in doing that until it's necessary.